### ISSUE TITLE
Question about validation metrics

### ISSUE THREAD

**Issue by TheLonelinessOfHS** [timeseriesAI/tsai#239]
Hi, sorry that this question is duplicate with a previous one. I am not quite familiar with fastai and it seems that the metrics values shown during the training are not from the validation set but from the training set. I am wondering if there is any approach to get the validation precision and recall score during the training? Or, is there any issue with my code below? Thanks.

```
    X, y = SlidingWindow(50, horizon=0, get_x=list(range(X.shape[1] - 1)), get_y=[X.shape[1] - 1])(X)
    y = y.astype(int)
    splits = TrainValidTestSplitter()(y[:-consts.TEST_SIZE])

    tfms  = [None, [TSClassification()]]
    dsets = TSDatasets(X[:-consts.TEST_SIZE], y[:-consts.TEST_SIZE], tfms=tfms, splits=splits)
    dls   = TSDataLoaders.from_dsets(dsets.train, dsets.valid, loss_func=nn.BCELoss(), bs=[64, 128], batch_tfms=[TSStandardize()], num_workers=0)
    test_ds = dls.valid.add_test(X[-consts.TEST_SIZE:], y[-consts.TEST_SIZE:]) # split like this because wish to test on a continuous set of time series data.
    test_dl = dls.valid.new(test_ds)

    model = InceptionTime(dls.vars, dls.c)
    learn = Learner(dls, model, metrics=[Precision(), Recall()])
    learn.fit_one_cycle(24, lr_max=1e-4)

    test_probas, test_targets, test_preds = learn.get_preds(dl=test_dl, with_decoded=True, save_preds=None, save_targs=None)
    print(f'accuracy: {skm.accuracy_score(test_targets, test_preds):10.6f}')
    print(f'precision: {skm.precision_score(test_targets, test_preds):10.6f}')
    print(f'recall: {skm.recall_score(test_targets, test_preds):10.6f}')

    test_probas, test_targets, test_preds = learn.get_preds(dl=dls.valid, with_decoded=True, save_preds=None, save_targs=None)
    print(f'validation accuracy: {skm.accuracy_score(test_targets, test_preds):10.6f}')
    print(f'validation precision: {skm.precision_score(test_targets, test_preds):10.6f}')
    print(f'validation recall: {skm.recall_score(test_targets, test_preds):10.6f}')
```

```
And the log shows me:
epoch     train_loss  valid_loss  precision_score  recall_score  time    
0         0.290075    0.283508    0.400000         0.000565      07:18                                                                                    
1         0.291655    0.283212    0.000000         0.000000      07:34                                                                                    
2         0.289442    0.281417    0.321839         0.007912      08:29                                                                                    
3         0.265994    0.272195    0.285714         0.007347      14:21                                                                                    
4         0.270511    0.274659    0.000000         0.000000      07:20                                                                                    
5         0.260311    0.257701    0.300000         0.000848      07:22                                                                                    
6         0.235657    0.254311    0.495495         0.093247      07:23                                                                                    
7         0.222522    0.237020    0.573529         0.055100      08:14                                                                                    
8         0.202300    0.218456    0.641837         0.177734      07:28                                                                                    
9         0.193323    0.231232    0.508783         0.466516      07:30                                                                                     
10        0.169027    0.194480    0.586676         0.462843      07:31                                                                                     
11        0.154449    0.182620    0.646642         0.437977      07:23                                                                                     
12        0.136467    0.174157    0.627341         0.539418      07:30                                                                                     
13        0.131988    0.158608    0.684532         0.537723      07:26                                                                                     
14        0.107252    0.153064    0.718612         0.549873      07:22                                                                                     
15        0.104701    0.151370    0.643562         0.718847      07:26                                                                                     
16        0.083210    0.145446    0.719638         0.629556      07:26                                                                                     
17        0.064595    0.137169    0.729333         0.645663      07:27                                                                                     
18        0.057357    0.137343    0.706613         0.691438      07:50                                                                                     
19        0.055130    0.136317    0.725439         0.701045      07:57                                                                                     
20        0.044213    0.137822    0.734011         0.693981      07:33                                                                                     
21        0.038672    0.136523    0.751800         0.678723      07:32                                                                                     
22        0.036667    0.136736    0.719174         0.718565      07:23                                                                                     
23        0.036205    0.135503    0.735816         0.703589      07:26                                                                                     
accuracy:   0.924300                                                                                                                      
precision:   0.125714
recall:   0.035144
validation accuracy:   0.913845                                                                                                                        
validation precision:   0.214286
validation recall:   0.000848
```

**Comment by oguiza** [comment]
Hi,
Sorry for my late response. 
It's actually the opposite. The metrics displayed during training come from the validation set. 
It's difficult to say if there's anything wrong with your code. Have you tried using CrossEntropyLossFlat instead of nn.BCELoss()? I don't know if y is binay, multiclass?

**Comment by oguiza** [comment]
I’ll close this issue due to a lack of a response. Please, reopen if necessary.
